design/open/high-level-python-wrapper.md

```diff
@@ -0,0 +1,314 @@
+    def modified_time(self) -> datetime.datetime: ...
+    def size_bytes(self) -> int: ...
+    def content_type(self) -> Optional[str]: ...
+    def metadata(self) -> dict[str, str]: ...
+    def physical_address(self) -> str: ...
+    def delete(self): ...
+
+
+class Change(NamedTuple):
+    type: Literal(['added', 'removed', 'changed', 'conflict,', 'prefix_changed'])
+    path: str
+    path_type: Literal(['common_prefix', 'object'])
+    size_bytes: Optional[int]
+
+
+class ServerConfiguration:
+    def __init__(self, client: Client = DefaultClient): ...
+    def version(self) -> str: ...
+    def storage_config(self) -> ServerStorageConfiguration: ...
+
+
+class ServerStorageConfiguration(NamedTuple):
+    blockstore_type: str
+    pre_sign_support: bool
+    import_support: bool
+
+
+class ImportManager:
+    def __init__(self, repository_id: str, reference_id: str, client: Client = DefaultClient): ...
+    def start(self) -> str:
+        'start import, reporting back (and storing) a process id'
+        ...
+    def wait(self, poll_interval: timedelta = timedelta(seconds=2)) -> ImportResult:
+        'poll a started import task ID, blocking until completion'
+        ...
+    def run(self, poll_interval: Optional[timedelta] = None) -> ImportResult:
+        'same as calling start() and then wait()'
+        ...
+
+
+class ImportResult(NamedTuple):
+    commit: Commit
+    ingested_objects: int
+
+class BranchTransaction(WriteIOBase):
+    def __init__(self, repository_id: str, source_branch_id: str, client: Client = DefaultClient): ...
+    
+    def __enter__(self):
+        'Create ephemeral branch from the source branch (e.g. <source_branch_id>-txn-<uuid>'
+        ...
+
+    def __exit__(self, type, value, traceback):
+        """if successful: commit, merge, delete ephemeral branch
+           otherwise, leave branch and report with a meaningful error
+        """
+        ...
+
+```
+
+While this list is fairly exhaustive, it might require a few additional tweaks and additions.
+
+Additionally, we define the following exception hierarchy:
+
+```python
+
+# lakefs.exceptions
+class LakeFSException(Exception):
+    status_code: int
+    message: str
+
+class NotFoundException(LakeFSException): ...
+class NotAuthorizedException(LakeFSException): ...
+class ServerException(LakeFSException): ...
+class UnsupportedOperationException(LakeFSException): ...
+```
+
+Other, more specific exceptions may subclass these, but all errors returned by the lakeFS server should sub-class one of these to make error handling easier for developers.
+
+The only exception should be errors returned by functions that, for compatibility should return specific error types. (this might be true for e.g. `open()` and errors returned by the `TextIO` or `BinaryIO` implementations).
+
+## Higher Level Utilities
+
+### I/O - reading/writing objects
+
+Provide a pythonic `open()` method that returns a "file-like object", passable to other frameworks or libraries
```

Comment:
As above, I would really prefer to separate open into open/create, with no mode argument.  I know that it is not Pythonic, but TBH the UN*X interface is pretty naff anyway for files, and even worse for objects.  We will confuse people, all of these are bad examples:

```py
with open(..., "w") as f:
    for r in f:
        # still gets here, no error, f thinks it's readable unless open returns a union type
        print(r)

with open(..., "a") as w:   # fails, open doesn't accept modes
    w.println("new last line")
open(..., "w+")  # fails, as do modes "r+", "a+".
open(..., "wt")  # <<< need to support this one!

Repository(repo).Ref('main').open(..., "w")    # succeeds(?)
Repository(repo).Ref('main~2').open(..., "w")  # method open exists but fails for "w"
Repository(repo).Ref('main~0').open(..., "w")  # fails, Repository.Ref is not referentially transparent
```

The return types on `open(..., "r")` and `open(..., "w")` are almost disjoint, merging them seems to me to be an ongoing major pain point for our users in exchange for a slightly easier ramp-up.

Reply:
Made `open()` read-only and will not accept a mode.